DR-3120:  Delete Azure Cloud Resources with Admin Option on Profile Delete

snf2ye opened the pull request:
The Changes:
- Added boolean option on the billing profile delete. When set to true, the associated azure storage account will be deleted. If there was a dataset/snapshot with secure monitoring enabled, then the log analytics and sentinel resources will be deleted as well. This option requires for a users to be an Admin in order to set it to true.
- Change for client (will be a breaking change when users upgrade their client): deleteProfile() will now require the the second boolean parameter to be included.
- Integration and test runner tests now clean up their azure storage account resources
- On dataset and snapshot delete, we now will set the marked_for_delete field on storage_account_resource entry in postgres to true instead of deleting the entry. They will now be deleted on billing profile delete.
- Updated a few unit tests to JUNIT5

snf2ye committed:
Add deleteCloudResources flag to delete profile endpoint

snf2ye committed:
Add Admin auth check; Wire in deleteCloudResources Flag

Move admin auth check to controller

merge conflicts: service, controller and controller test

spotless

spotless

snf2ye committed:
Convert ProfileServiceUnitTest to JUnit5

merge conflicts: service unit test

snf2ye committed:
deleteCloudResources correctly passed to job builder in profile service

snf2ye committed:
Convert ProfileAPIControllerTest to JUNIT5

snf2ye committed:
Assert correct auth checks are made for delete billing profile

snf2ye committed:
Add sentinel and log workspace delete steps to flight

cleanup javadoc

spotless

Wire in delete monitoring steps to delete profile flight

snf2ye committed:
Only delete storage accounts marked for delete

snf2ye committed:
Add delete storage account step

snf2ye committed:
Add Step

snf2ye committed:
Add profile delete with resources to Azure integration test

snf2ye committed:
Mark storage accounts for delete rather than deleting database entry

snf2ye commented on src/main/java/bio/terra/service/resourcemanagement/ResourceService.java:
```diff
@@ -426,7 +426,7 @@ public void deleteStorageContainer(UUID storageResourceId, UUID profileId, Strin
     azureContainerPdao.deleteContainer(snapshotBillingProfile, storageAccountResource);
 
-    storageAccountService.deleteCloudStorageAccountMetadata(
+    storageAccountService.markForDeleteCloudStorageAccountMetadata(
```
Note to reviewers: This is a switch from the original plan. Instead of deleting the entry in `storage_account_resource` on dataset and snapshot delete, I'm only marking them for delete. They will also be deleted on billing profile delete (whether or not the admin "cloud resource delete" flag is set)

snf2ye committed:
Fix making parameter optional

snf2ye committed:
Remove deleteCloudStorageAccountMetadata and add test

spotless

Delete cloud resources

Delete cloud resources in test runner test

again

pshapiro4broad commented on src/main/java/bio/terra/service/profile/ProfileApiController.java:
```diff
@@ -173,4 +180,12 @@ private void verifyProfileAuthorization(
   }
+
+  private void verifyAdminAuthorization(AuthenticatedUserRequest userReq) {
+    iamService.verifyAuthorization(
+        userReq,
+        IamResourceType.DATAREPO,
+        applicationConfiguration.getResourceId(),
+        IamAction.CONFIGURE);
```
Could/should this be a general method in `IamService`?

snf2ye replied:
I think we would then need to wire in ApplicationConfiguration to IamService to get the resource Id. Given that we only do this admin check in a few places, would it be worth pull through applicationConfiguration everywhere that we use call iamservice? But, maybe I'm misunderstanding what would be required!

pshapiro4broad replied:
Yes, moving it into IamService would require wiring ApplicationConfiguration into it. It wouldn't change the way IamService APIs are called but it would add a new dependency so it might not be worth doing.

The thing I was thinking was that TDR uses a check on the `CONFIGURE` action as a proxy for an admin user check, and that's the thing I wanted to clean up. But now that I look at uses of `ApplicationConfiguration.getResourceId()` I see that there's a couple different actions that are checked. Maybe this is a ticket for a future cleanup; it seems like if we're using `CONFIGURE` as a proxy for `is admin user`, it should really be `ADMIN` or something else. Either that, or we should have different actions for different admin type operations, such as `SYNC_DUOS_USERS`.

snf2ye replied:
Checking the "configure" action as a proxy for the admin check also provides a confusing error message for the user. (Rather than saying you need to be an admin to complete the action, it just says that your user doesn't have the 'configure' action). 
Ticket entered here: https://broadworkbench.atlassian.net/browse/DR-3248

pshapiro4broad commented on src/main/java/bio/terra/service/profile/ProfileService.java:
```diff
@@ -122,10 +122,13 @@ public String updateProfile(
    */
-  public String deleteProfile(UUID id, AuthenticatedUserRequest user) {
+  public String deleteProfile(
+      UUID id, Boolean deleteCloudResources, AuthenticatedUserRequest user) {
```
At this point, we know that `deleteCloudResources` is non-null, so it should be declared as `boolean`.
```suggestion
      UUID id, boolean deleteCloudResources, AuthenticatedUserRequest user) {
```

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureMonitoringService.java:
```diff
@@ -143,6 +143,31 @@ public void deleteLogAnalyticsWorkspace(BillingProfileModel profileModel, String
+      BillingProfileModel profileModel, AzureStorageAccountResource storageAccountResource) {
+    LogAnalyticsManager client =
+        resourceConfiguration.getLogAnalyticsManagerClient(
+            resourceConfiguration.credentials().getHomeTenantId(),
```
It looks like every time `getLogAnalyticsManagerClient()` is called, the caller passes in `resourceConfiguration.credentials().getHomeTenantId()` as the first argument. Could this be done inside the method instead? It looks like the same thing is true of `getSecurityInsightsManagerClient()` too.

snf2ye replied:
Oh yeah, that makes a lot of sense!

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureStorageAccountService.java:
```diff
@@ -205,16 +208,47 @@ public AzureStorageAccountResource retrieveStorageAccountById(UUID storageAccoun
+      List<UUID> applicationResourceIds, boolean markedForDelete) {
+    List<AzureStorageAccountResource> resources = new ArrayList<>();
+    applicationResourceIds.stream()
+        .forEach(
```
You can stream the result directly into a `List` by using `flatMap()` here instead of `forEach()`. This should work:
```
    return applicationResourceIds.stream()
        .flatMap(
            applicationResourceId ->
                resourceDao.retrieveStorageAccountsByApplicationResource(
                    applicationResourceId, markedForDelete).stream()).toList();
```

snf2ye replied:
Interesting - so, this change caused the serialization into the flight map to break. The result of this method is directly put in the flight map with type List<AzureStroageAccountResource>. I'm guessing the issue is that we're not directly initializing the array list?

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/flight/AzureStorageMonitoringStepProvider.java:
```diff
@@ -65,5 +66,15 @@ public List<StepDef> configureSteps(boolean isSecureMonitoringEnabled, AzureRegi
+        new StepDef(
+            new DeleteSentinelStep(monitoringService, azureStorageAccountService), noRetry));
+    steps.add(new StepDef(new DeleteLogAnalyticsWorkspaceStep(monitoringService), noRetry));
+    return steps;
```
If the list of steps isn't dynamic you could return a `List.of()` here
```
    return List.of(new StepDef(...), new StepDef(..));
```

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/flight/AzureStorageMonitoringStepProvider.java:
```diff
@@ -65,5 +66,15 @@ public List<StepDef> configureSteps(boolean isSecureMonitoringEnabled, AzureRegi
+    return steps;
+  }
+
   public record StepDef(Step step, RetryRule retryRule) {}
```
It looks like a common use case is to provide a retry of `non` for a step. Adding a custom constructor to the record for that case could avoid the extra code that declares and passes the `noRetry` argument.
```
  public record StepDef(Step step, RetryRule retryRule) {
    public StepDef(Step step) {
      this(step, getRetryRuleNone());
    }
  }
```

snf2ye replied:
Oh nice, that's a great idea.

snf2ye commented on datarepo-clienttests/src/main/java/scripts/utils/tdrwrapper/DataRepoWrap.java:
```diff
@@ -173,7 +173,7 @@ public DeleteResponseModel deleteProfile(UUID profileId) throws Exception {
 
   public WrapFuture<DeleteResponseModel> deleteProfileFuture(UUID profileId) {
-    JobModel jobResponse = apiCallThrow(() -> resourcesApi.deleteProfile(profileId));
+    JobModel jobResponse = apiCallThrow(() -> resourcesApi.deleteProfile(profileId, true));
```
We want to delete cloud resources in all of our tests, but this is an example that shows that adding this parameter is a breaking change for those using the client: It requires that the second boolean parameter is provided. 
I've set this parameter to be optional in the open api spec, but it appears that with the java client it still acts as a required parameter (I've found references to this 'bug' online)

snf2ye committed:
Fix testRunner: Add second parameter

Revert "Fix testRunner: Add second parameter"

This reverts commit 88ac6fd744cf80490a133790792d076cb45d4735.

Use admin for azure related test runner test

update

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/flight/DeleteSentinelStep.java:
```diff
@@ -0,0 +1,54 @@
+        logger.info(
+            "Sentinel instance found for storage account {}; Attempting delete.",
+            storageAccountResource.getName());
+        monitoringService.deleteSentinelByStorageAccount(profileModel, storageAccountResource);
```
Maybe this is just the way the azure APIs work, but it seems odd to me that all three cases are `getX() != null` and then `deleteX()`, but you don't pass the result of the `getX()` operation to the delete operation.

snf2ye replied:
Oh, no, I don't think it has to follow this pattern, but I do think it makes sense in this case. 

It does follow the pattern of the create step -- We check if the resource already exists before trying to create it. 
But, more than following this pattern, I think this is just an alternate to wrapping the delete in a try/catch. While this approach doubles the number of requests to Azure, it should help us debug issues with the delete. We'll know if we're attempting the delete then the resource should exist to be deleted. I think this is a fair trade because this won't be a heavily trafficked endpoint, but any information we can provide on it's success/failure will be helpful.

pshapiro4broad replied:
Sorry I wasn't clear, I think the check before delete is fine (vs catch when delete fails), the part in particular I was wondering about is why you can't pass the result of `get()` to the `delete()` API. E.g., it looks like you could pass in the `id` from the object you get back:
```
        monitoringService.deleteSentinelById(profileModel, sentinel.id());
```

Also I wonder if the `get` should be baked into the `delete` or not; do we expect that every caller will need to check before calling `delete`? the only caller I could find of `deleteSentinelById()` is in `CreateSentinelStep.undoStep()`, which doesn't call `get` first, but the lack of calling `get` looks like a bug, since a step's undo should be idempotent.

pshapiro4broad commented on src/test/java/bio/terra/common/fixtures/ConnectedOperations.java:
```diff
@@ -486,7 +486,9 @@ public boolean deleteTestDataset(UUID id) throws Exception {
   public boolean deleteTestProfile(UUID id) throws Exception {
-    MvcResult result = mvc.perform(delete("/api/resources/v1/profiles/" + id)).andReturn();
+    MvcResult result =
+        mvc.perform(delete("/api/resources/v1/profiles/" + id + "?deleteCloudResources=true"))
```
The mvc APIs provide templating so you don't need to use string concatenation.
```suggestion
        mvc.perform(delete("/api/resources/v1/profiles/{id}?deleteCloudResources=true", id))
```

pshapiro4broad commented on src/test/java/bio/terra/integration/DataRepoClient.java:
```diff
@@ -105,7 +105,18 @@ public <T> DataRepoResponse<T> put(
+      entity = new HttpEntity<>(json, getHeaders(user));
+    } else {
+      entity = new HttpEntity<>(getHeaders(user));
+    }
```
The first argument to `HttpEntity(body, headers)` is nullable so you don't need an `if` here.
```suggestion
    HttpEntity<String> entity = new HttpEntity<>(json, getHeaders(user));
```

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileAPIControllerTest.java:
```diff
@@ -32,20 +35,24 @@
-@RunWith(MockitoJUnitRunner.StrictStubs.class)
 @ActiveProfiles({"google", "unittest"})
-@Category(Unit.class)
+@ExtendWith(MockitoExtension.class)
```
Now that this is "really" using strict stubs, you can safely remove many of the `verify()` calls here. For example, the  `verify()` call in `testCreateProfile()` can be removed. Any API that's explicitly mocked and is being checked for `times(1)` will automatically be verified when the test completes.

snf2ye replied:
Ok, great, I've removed all of the verify checks that are explicitly mocked.

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileAPIControllerTest.java:
```diff
@@ -141,44 +150,66 @@ public void testUpdateProfileForbidden() {
     UUID deleteId = UUID.fromString("aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee");
     String jobId = "jobId";
-    when(profileService.deleteProfile(eq(deleteId), eq(user))).thenReturn(jobId);
+    when(profileService.deleteProfile(eq(deleteId), eq(deleteCloudResources), eq(user)))
+        .thenReturn(jobId);
```
if all parameters to a mock are constant, you don't need to use `eq()`.
```suggestion
    when(profileService.deleteProfile(deleteId, deleteCloudResources, user)).thenReturn(jobId);
```

snf2ye replied:
Ah, ok - You don't need it when mocking but do on verify.

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileAPIControllerTest.java:
```diff
@@ -141,44 +150,66 @@ public void testUpdateProfileForbidden() {
 
     var jobModel = new JobModel();
     jobModel.setJobStatus(JobStatusEnum.RUNNING);
     when(jobService.retrieveJob(eq(jobId), eq(user))).thenReturn(jobModel);
```
```suggestion
    when(jobService.retrieveJob(jobId, user)).thenReturn(jobModel);
```

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileAPIControllerTest.java:
```diff
@@ -141,44 +150,66 @@ public void testUpdateProfileForbidden() {
 
-    ResponseEntity entity = apiController.deleteProfile(deleteId);
-    verify(profileService, times(1)).deleteProfile(eq(deleteId), eq(user));
+    ResponseEntity entity = apiController.deleteProfile(deleteId, deleteCloudResources);
```
In general we want to avoid the use of raw generic (parameterized) types when possible. In this case you can use the correct type:
```
    ResponseEntity<JobModel> entity = apiController.deleteProfile(deleteId, deleteCloudResources);
```

or you could use `var`
```
    var entity = apiController.deleteProfile(deleteId, deleteCloudResources);
```

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileAPIControllerTest.java:
```diff
@@ -141,44 +150,66 @@ public void testUpdateProfileForbidden() {
+            eq(IamAction.DELETE));
+    verify(profileService, times(1))
+        .deleteProfile(eq(deleteId), eq(deleteCloudResources), eq(user));
     assertNotNull(entity);
```
Since the mock is returning a valid job model, you could assert that here.
```
    MatcherAssert.assertThat(entity.getBody(), Matchers.is(jobModel));
```
(I would statically import `assertThat()` and `is()` but I included the classes here so it's clear which ones I'm referring to)

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileAPIControllerTest.java:
```diff
@@ -141,44 +150,66 @@ public void testUpdateProfileForbidden() {
+        .thenReturn(jobId);
+    when(profileService.getProfileByIdNoCheck(deleteId))
+        .thenReturn(new BillingProfileModel().id(deleteId));
+    doNothing().when(iamService).verifyAuthorization(any(), any(), any(), any());
```
A (default) mocked class has a "do nothing" implementation for each member, so you should be able to remove this line without changing the test behavior.

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileAPIControllerTest.java:
```diff
@@ -141,44 +150,66 @@ public void testUpdateProfileForbidden() {
+        .verifyAuthorization(
+            eq(user), eq(IamResourceType.DATAREPO), any(), eq(IamAction.CONFIGURE));
+    // Always check that the user has access on the spend profile
+    verify(iamService, times(1))
+        .verifyAuthorization(
+            eq(user),
+            eq(IamResourceType.SPEND_PROFILE),
+            eq(deleteId.toString()),
+            eq(IamAction.DELETE));
```
As before, the arguments are all literals so `eq()` is not needed. Also `times(1)` is the default so that can be omitted.
```suggestion
    verify(iamService).verifyAuthorization(user,IamResourceType.SPEND_PROFILE, deleteId.toString(), IamAction.DELETE);
```

snf2ye replied:
Oh, ok, all the arguments must be constant to remove it. Thanks for that clarification!

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileAPIControllerTest.java:
```diff
@@ -141,44 +150,66 @@ public void testUpdateProfileForbidden() {
+            eq(IamResourceType.SPEND_PROFILE),
+            eq(deleteId.toString()),
+            eq(IamAction.DELETE));
+    verify(profileService, times(1))
+        .deleteProfile(eq(deleteId), eq(deleteCloudResources), eq(user));
```
Since this API is explicitly mocked, it will be verified by strict stubs and this verify call isn't needed.

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileAPIControllerTest.java:
```diff
@@ -141,44 +150,66 @@ public void testUpdateProfileForbidden() {
+        ProfileNotFoundException.class, () -> apiController.deleteProfile(profileId, false));
     verifyNoInteractions(iamService);
-    verify(profileService, times(0)).deleteProfile(eq(profileId), eq(user));
+    verify(profileService, times(0)).deleteProfile(eq(profileId), eq(false), eq(user));
```
Instead of `times(0)` you can write `never()`. Also these are all literal arguments, so `eq` can be removed.
```suggestion
    verify(profileService, never()).deleteProfile(profileId, false, user);
```

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileAPIControllerTest.java:
```diff
@@ -32,20 +35,24 @@
-@Category(Unit.class)
+@ExtendWith(MockitoExtension.class)
+@Tag(Unit.TAG)
 public class ProfileAPIControllerTest {
```
it's probably too big a change to add here, but since this is a controller test I would expect it to use a mock mvc API and call the endpoints directly instead of invoking them via method calls.

snf2ye replied:
Oh yeah, that is a good point. Yeah, I think this is out of scope of this change, but definitely something to do in the future.

pshapiro4broad commented on src/test/java/bio/terra/service/profile/ProfileServiceUnitTest.java:
```diff
@@ -133,6 +138,9 @@ public void testDeleteProfile() {
     when(jobBuilder.addParameter(eq(JobMapKeys.IAM_ACTION.getKeyName()), eq(IamAction.DELETE)))
         .thenReturn(jobBuilder);
+    when(jobBuilder.addParameter(
+            eq(JobMapKeys.DELETE_CLOUD_RESOURCES.getKeyName()), eq(deleteCloudResources)))
```
```suggestion
            JobMapKeys.DELETE_CLOUD_RESOURCES.getKeyName(), deleteCloudResources))
```

pshapiro4broad commented on src/test/java/bio/terra/service/resourcemanagement/azure/AzureResourceDaoTest.java:
```diff
@@ -139,4 +139,32 @@ public void oneApplicationOneBillingProfilesTwoStorageAccounts() throws Exceptio
+  @Test
+  public void markForDeleteStorageAccountMetadata() {
+    var appDeploymentId = applicationDeployments.get(0).getId();
+    assertThat(
+        "Before marking for delete, confirm that 2 storage accounts are returned when marked_for_delete = false",
+        azureResourceDao
+            .retrieveStorageAccountsByApplicationResource(appDeploymentId, false)
+            .size(),
+        equalTo(2));
```
There is a built-in matcher for collection size, `hasSize()`.
```suggestion
    assertThat(
        "Before marking for delete, confirm that 2 storage accounts are returned when marked_for_delete = false",
        azureResourceDao
            .retrieveStorageAccountsByApplicationResource(appDeploymentId, false),
        hasSize(2));
```

snf2ye replied:
Oh right, thanks!

pshapiro4broad commented on src/test/java/bio/terra/service/resourcemanagement/azure/AzureResourceDaoTest.java:
```diff
@@ -139,4 +139,32 @@ public void oneApplicationOneBillingProfilesTwoStorageAccounts() throws Exceptio
+
+    var storageAccountsMarkedForDelete =
+        azureResourceDao.retrieveStorageAccountsByApplicationResource(appDeploymentId, true);
+    assertThat(
+        "Only 1 storage accounts is returned", storageAccountsMarkedForDelete.size(), equalTo(1));
```
```suggestion
    assertThat("Only 1 storage accounts is returned", storageAccountsMarkedForDelete, hasSize(1));
```

pshapiro4broad reviewed (commented):
I have a few comments, many about the tests which are copies of existing code which I think we can improve on.

snf2ye committed:
Check either that user is admin or has access on profile

- Fix test to no longer check spend profile auth if admin check is being performed

snf2ye committed:
Code Review

- Use hasSize matcher
- boolean instead of Boolean

snf2ye commented on src/main/java/bio/terra/service/resourcemanagement/flight/RecordAzureStorageAccountsStep.java:
```diff
@@ -0,0 +1,46 @@
+    List<UUID> appIdList =
+        workingMap.get(ProfileMapKeys.PROFILE_APPLICATION_DEPLOYMENT_ID_LIST, List.class);
+
+    // Note: This will list a storage account for each top level container
+    // So, there may be duplicates of the cloud storage account resource
+    // Top level container and storage accounts have a many-to-one relationship
+    List<AzureStorageAccountResource> storageAccounts =
+        azureStorageAccountService.listStorageAccountPerAppDeployment(appIdList, true);
+    workingMap.put(ProfileMapKeys.PROFILE_STORAGE_ACCOUNT_RESOURCE_LIST, storageAccounts);
```
Reviewers: I would appreciate some feedback on this issue/approach.

nmalfroy replied:
I think this makes sense.  I have a ticket to refactor the storage of these things but yeah, you could have this go across multiple storage accounts,  not just top level containers (e.g. when you have different regions).  I left another comment about distinct-ing down to just the storage account (since this list might get kinda big if there are many datasets...though as I type this, I realize that you would have to delete all datasets/snapshots before actually being able to delete so the list wouldn't be big)

snf2ye replied:
There would still be entry for each dataset/snapshot since we're just marking the stroage_account_resource for delete on dataset/snapshot delete. 

I think I'll go ahead and make this list distinct and just make it clear in the profile map key that it's only distinct storage accounts. I'm just worried in the future we could think we have a list of all of the top level containers, but we would lose that if we just get the distinct storage accounts.

snf2ye committed:
code review: remove eq() from test

snf2ye committed:
getHomeTenantId in Service

snf2ye committed:
Code Review: Update test to use never()

snf2ye committed:
Code Review: Update list logc, add test

snf2ye committed:
Code review: Use List.of for step list

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureMonitoringService.java:
```diff
@@ -74,9 +74,7 @@ public AzureMonitoringService(AzureResourceConfiguration resourceConfiguration)
     LogAnalyticsManager client =
-        resourceConfiguration.getLogAnalyticsManagerClient(
-            resourceConfiguration.credentials().getHomeTenantId(),
-            profileModel.getSubscriptionId());
```
Nice refactoring!

snf2ye replied:
Agreed! Thanks for catching that - this is much cleaner.

pshapiro4broad reviewed (approved):
Looks good to me

snf2ye committed:
No retry option on StepDef

snf2ye committed:
Code Review: String concat

snf2ye committed:
Code Review: Json is nullable

snf2ye committed:
Code Review: Remove verify when method already mocked; not needed when. using strict stubbing

snf2ye committed:
Code Review: review eq() from mocks

snf2ye committed:
Code Review: Use correct type of response in test

snf2ye committed:
Code Review: ACtually check the contents of the job model returned in test; remove doNothing mock call

snf2ye committed:
Code Review: Remove eq() from verify arguments when all constant

nmalfroy commented on src/main/java/bio/terra/service/resourcemanagement/flight/DeleteLogAnalyticsWorkspaceStep.java:
```diff
@@ -0,0 +1,51 @@
+            "Log Analytics Workspace found for storage account {}; Attempting delete.",
+            storageAccountResource.getName());
+        // TODO - how should we do error handling?
+        monitoringService.deleteLogAnalyticsWorkspace(profileModel, storageAccountResource);
```
I think this is ok for now. Again, this is a super rare operation so we could go into clean up if we have to

nmalfroy commented on src/main/java/bio/terra/service/resourcemanagement/flight/DeleteAzureStorageAccountStep.java:
```diff
@@ -0,0 +1,48 @@
+        workingMap.get(
+            ProfileMapKeys.PROFILE_STORAGE_ACCOUNT_RESOURCE_LIST, new TypeReference<>() {});
+
+    for (AzureStorageAccountResource storageAccountResource : storageAccounts) {
```
This is a weirdness with how I laid out the tables but this will lead to you trying to delete the same storage account for each of its top level containers.  You might want to have this to a distinct of the storage accounts based on the storage account name and then change the `azureStorageAccountService.deleteCloudStorageAccount(BillingProfileModel, AzureStorageAccountResource)` method to azureStorageAccountService.deleteCloudStorageAccount(BillingProfileModel, String appId, String storageAccountName)`

Just thinking of a case like AnVIL where it would iterate over this hundreds of times

nmalfroy reviewed (approved):
I left a comment about the potentially large number of storage account deletes but I think I've convinced myself that it's not a big deal since users have to make sure all datasets and snapshots are deleted before you can delete an billing profile.  I left the comments just in case you think it might be an issue but I think this is good to merge

snf2ye committed:
Sonar code smells fixed

pshapiro4broad commented on src/main/java/bio/terra/service/profile/flight/delete/ProfileDeleteFlight.java:
```diff
@@ -33,6 +41,9 @@ public ProfileDeleteFlight(FlightMap inputParameters, Object applicationContext)
         CloudPlatformWrapper.of(
             inputParameters.get(JobMapKeys.CLOUD_PLATFORM.getKeyName(), String.class));
 
+    AzureStorageMonitoringStepProvider azureStorageMonitoringStepProvider =
+        new AzureStorageMonitoringStepProvider(monitoringService);
+
```
If this is only used inside the `if` below, then it could be defined there instead. That limits its scope and prevents unneeded code from running.

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/flight/DeleteLogAnalyticsWorkspaceStep.java:
```diff
@@ -0,0 +1,50 @@
+public class DeleteLogAnalyticsWorkspaceStep extends DefaultUndoStep {
+  private static final Logger logger =
+      LoggerFactory.getLogger(DeleteLogAnalyticsWorkspaceStep.class);
+  private AzureMonitoringService monitoringService;
```
field can be `final`
```suggestion
  private final AzureMonitoringService monitoringService;
```

snf2ye committed:
Revert code review change -- the other route broke the serialization in the map

snf2ye committed:
Code Review: Move monitoring provider; add final

snf2ye committed:
Filter down to unique storage accounts

remove space

snf2ye commented on src/main/java/bio/terra/service/resourcemanagement/flight/RecordAzureStorageAccountsStep.java:
```diff
@@ -43,4 +42,20 @@ public StepResult doStep(FlightContext context) throws InterruptedException, Ret
+    List<AzureStorageAccountResource> storageAccounts =
+        azureStorageAccountService.listStorageAccountPerAppDeployment(appIdList, true);
+    // Filter down this list to only return one resource per unique cloud storage account resource
+    return storageAccounts.stream().filter(distinctByKey(sa -> sa.getName())).toList();
```
@nmalfroy  - Is filtering down on the storage account name the right move here? I think it's the only field that's unique to the storage account -- the id is unique per storage account/top level container combination.

snf2ye committed:
Serialization was mad again - Adding resource to an array helps.
